modules/ml/src/main/java/org/apache/ignite/ml/trainers/TrainerTransformers.java

```diff
@@ -0,0 +1,373 @@
+
+        // If we need to do projection, do it.
+        if (mappings != null) {
+            for (int i = 0; i < models.size(); i++) {
+                models.get(i).setMapping(getProjector(mappings.get(i)));
+            }
+        }
+
+        double learningTime = (double)(System.currentTimeMillis() - startTs) / 1000.0;
+        log.log(MLLogger.VerboseLevel.LOW, "The training time was %.2fs.", learningTime);
+        log.log(MLLogger.VerboseLevel.LOW, "Learning finished.");
+
+        return new ModelsComposition(models, aggregator);
+    }
+
+    /**
+     * Get mapping R^featuresVectorSize -> R^maximumFeaturesCntPerMdl.
+     *
+     * @param featuresVectorSize Features vector size (Dimension of initial space).
+     * @param maximumFeaturesCntPerMdl Dimension of target space.
+     * @param seed Seed.
+     * @return Mapping R^featuresVectorSize -> R^maximumFeaturesCntPerMdl.
+     */
+    public static int[] getMapping(int featuresVectorSize, int maximumFeaturesCntPerMdl, long seed) {
+        return Utils.selectKDistinct(featuresVectorSize, maximumFeaturesCntPerMdl, new Random(seed));
+    }
+
+    /**
+     * Get projector from index mapping.
+     *
+     * @param mapping Index mapping.
+     * @return Projector.
+     */
+    public static IgniteFunction<Vector, Vector> getProjector(int[] mapping) {
+        return v -> {
+            Vector res = VectorUtils.zeroes(mapping.length);
+            for (int i = 0; i < mapping.length; i++) {
+                res.set(i, v.get(mapping[i]));
+            }
+            return res;
+        };
+    }
+
+    /**
+     * Creates feature extractor which is a composition of given feature extractor and projection given by
+     * coordinate indexes mapping.
+     *
+     * @param featureExtractor Initial feature extractor.
+     * @param featureMapping Coordinate indexes mapping.
+     * @param <K> Type of keys.
+     * @param <V> Type of values.
+     * @return Composition of given feature extractor and projection given by coordinate indexes mapping.
+     */
+    private static <K, V> IgniteBiFunction<K, V, Vector> wrapExtractor(IgniteBiFunction<K, V, Vector> featureExtractor,
+        int[] featureMapping) {
+        return featureExtractor.andThen((IgniteFunction<Vector, Vector>)featureValues -> {
+            double[] newFeaturesValues = new double[featureMapping.length];
+            for (int i = 0; i < featureMapping.length; i++) {
+                newFeaturesValues[i] = featureValues.get(featureMapping[i]);
+            }
+            return VectorUtils.of(newFeaturesValues);
+        });
+    }
+
+    /**
+     * Model with mapping from X to X.
+     *
+     * @param <X> Input space.
+     * @param <Y> Output space.
+     * @param <M> Model.
+     */
+    private static class ModelWithMapping<X, Y, M extends Model<X, Y>> implements Model<X, Y> {
+        /** Model. */
+        private M model;
+
+        /** Mapping. */
+        private IgniteFunction<X, X> mapping;
```

Comment:
Can we make these fields final?

Reply:
"Yes" for model, "no" for mappings. Done for model.